File: backend/internal/parser/parser.go
// internal/parser/parser.go - Parser implementation

package parser

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"

	"nula-backend/internal/ast"
)

type Parser struct {
	tokens []Token
	pos    int
}

type Token struct {
	Type  TokenType
	Value string
}

type TokenType int

const (
	TokenIdent TokenType = iota
	TokenNumber
	TokenString
	TokenOperator
	TokenKeyword
	TokenSymbol
	TokenEof
)

var keywords = map[string]bool{
	"if":    true,
	"else":  true,
	"while": true,
	"for":   true,
	"fn":    true,
	"var":   true,
	"write": true,
}

func tokenize(code string) []Token {
	var tokens []Token
	lines := strings.Split(code, "\n")
	inMultiComment := false

	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" {
			continue
		}

		if inMultiComment {
			if strings.Contains(trimmed, "!") {
				inMultiComment = false
			}
			continue
		}

		if strings.HasPrefix(trimmed, "!") {
			inMultiComment = true
			continue
		}

		if strings.HasPrefix(trimmed, "@") {
			continue
		}

		// Tokenize line
		i := 0
		for i < len(trimmed) {
			c := trimmed[i]
			if isLetter(c) {
				id := ""
				for i < len(trimmed) && (isLetter(trimmed[i]) || isDigit(trimmed[i]) || trimmed[i] == '_') {
					id += string(trimmed[i])
					i++
				}
				if keywords[id] {
					tokens = append(tokens, Token{TokenKeyword, id})
				} else {
					tokens = append(tokens, Token{TokenIdent, id})
				}
				continue
			}
			if isDigit(c) || (c == '.' && i+1 < len(trimmed) && isDigit(trimmed[i+1])) {
				num := ""
				for i < len(trimmed) && (isDigit(trimmed[i]) || trimmed[i] == '.') {
					num += string(trimmed[i])
					i++
				}
				tokens = append(tokens, Token{TokenNumber, num})
				continue
			}
			if c == '"' {
				i++
				str := ""
				for i < len(trimmed) && trimmed[i] != '"' {
					str += string(trimmed[i])
					i++
				}
				i++ // skip closing "
				tokens = append(tokens, Token{TokenString, str})
				continue
			}
			if strings.Contains("+-*/^=<>!&|", string(c)) {
				tokens = append(tokens, Token{TokenOperator, string(c)})
				i++
				continue
			}
			if strings.Contains("(){}[]:;,", string(c)) {
				tokens = append(tokens, Token{TokenSymbol, string(c)})
				i++
				continue
			}
			if c == ':' && i+1 < len(trimmed) && trimmed[i+1] == ':' {
				importName := strings.TrimPrefix(trimmed[i+2:], ":")
				tokens = append(tokens, Token{TokenKeyword, "import"})
				tokens = append(tokens, Token{TokenIdent, importName})
				break // Assume whole line
			}
			if c == '<' && strings.HasSuffix(trimmed, ">") {
				module := strings.Trim(trimmed, "<>")
				tokens = append(tokens, Token{TokenKeyword, "from"})
				tokens = append(tokens, Token{TokenIdent, module})
				break
			}
			embeddedRe := regexp.MustCompile(`# =(\w+)= \[(.*)\]`)
			if match := embeddedRe.FindStringSubmatch(trimmed); match != nil {
				tokens = append(tokens, Token{TokenKeyword, "embedded"})
				tokens = append(tokens, Token{TokenIdent, match[1]})
				tokens = append(tokens, Token{TokenString, match[2]})
				break
			}
			i++
		}
	}
	tokens = append(tokens, Token{TokenEof, ""})
	return tokens
}

func Parse(code string) (ast.Program, error) {
	p := &Parser{tokens: tokenize(code)}
	return ast.Program{Statements: p.parseProgram()}, nil
}

func (p *Parser) parseProgram() []ast.Node {
	var stmts []ast.Node
	for p.current() != TokenEof {
		stmts = append(stmts, p.parseStmt())
	}
	return stmts
}

func (p *Parser) parseStmt() ast.Node {
	switch p.currentType() {
	case TokenKeyword:
		switch p.currentValue() {
		case "var":
			return p.parseVarDecl()
		case "if":
			return p.parseIf()
		case "while":
			return p.parseWhile()
		case "for":
			return p.parseFor()
		case "fn":
			return p.parseFuncDef()
		case "write":
			return p.parseWrite()
		case "import":
			return p.parseImport()
		case "from":
			return p.parseFrom()
		case "embedded":
			return p.parseEmbedded()
		}
	case TokenIdent:
		return p.parseAssignOrCall()
	}
	return p.parseExpr()
}

func (p *Parser) parseVarDecl() ast.Node {
	p.advance() // var
	name := p.expect(TokenIdent).Value
	p.expectOp("=")
	value := p.parseExpr()
	return &ast.VarDecl{Name: name, Value: value}
}

func (p *Parser) parseAssignOrCall() ast.Node {
	name := p.expect(TokenIdent).Value
	if p.currentValue() == "=" {
		p.advance()
		return &ast.Assign{Name: name, Value: p.parseExpr()}
	}
	if p.currentValue() == "(" {
		return &ast.FuncCall{Name: name, Args: p.parseArgs()}
	}
	return &ast.Var{Name: name}
}

func (p *Parser) parseIf() ast.Node {
	p.advance() // if
	cond := p.parseExpr()
	p.expectSym("{")
	then := p.parseBlock()
	p.expectSym("}")
	var els []ast.Node
	if p.currentValue() == "else" {
		p.advance()
		p.expectSym("{")
		els = p.parseBlock()
		p.expectSym("}")
	}
	return &ast.IfStmt{Condition: cond, Then: then, Else: els}
}

func (p *Parser) parseWhile() ast.Node {
	p.advance() // while
	cond := p.parseExpr()
	p.expectSym("{")
	body := p.parseBlock()
	p.expectSym("}")
	return &ast.WhileStmt{Condition: cond, Body: body}
}

func (p *Parser) parseFor() ast.Node {
	p.advance() // for
	varName := p.expect(TokenIdent).Value
	p.expectKeyword("in")
	start := p.parseExpr()
	p.expectOp("..")
	end := p.parseExpr()
	p.expectSym("{")
	body := p.parseBlock()
	p.expectSym("}")
	return &ast.ForStmt{Var: varName, Start: start, End: end, Body: body}
}

func (p *Parser) parseFuncDef() ast.Node {
	p.advance() // fn
	name := p.expect(TokenIdent).Value
	p.expectSym("(")
	var params []string
	for p.current() != TokenSymbol || p.currentValue() != ")" {
		params = append(params, p.expect(TokenIdent).Value)
		if p.currentValue() == "," {
			p.advance()
		}
	}
	p.expectSym(")")
	p.expectSym("{")
	body := p.parseBlock()
	p.expectSym("}")
	return &ast.FuncDef{Name: name, Params: params, Body: body}
}

func (p *Parser) parseWrite() ast.Node {
	p.advance() // write
	return &ast.WriteStmt{Expr: p.parseExpr()}
}

func (p *Parser) parseImport() ast.Node {
	p.advance() // import
	name := p.expect(TokenIdent).Value
	return &ast.ImportStmt{Name: name}
}

func (p *Parser) parseFrom() ast.Node {
	p.advance() // from
	name := p.expect(TokenIdent).Value
	// For simplicity, treat as import
	return &ast.ImportStmt{Name: name}
}

func (p *Parser) parseEmbedded() ast.Node {
	p.advance() // embedded
	lang := p.expect(TokenIdent).Value
	code := p.expect(TokenString).Value
	return &ast.Embedded{Lang: lang, Code: code}
}

func (p *Parser) parseBlock() []ast.Node {
	var block []ast.Node
	for p.currentValue() != "}" && p.current() != TokenEof {
		block = append(block, p.parseStmt())
	}
	return block
}

func (p *Parser) parseExpr() ast.Node {
	return p.parseAdd()
}

func (p *Parser) parseAdd() ast.Node {
	left := p.parseMul()
	for p.isOp("+") || p.isOp("-") {
		op := p.advance().Value
		right := p.parseMul()
		left = &ast.BinOp{Op: op, Left: left, Right: right}
	}
	return left
}

func (p *Parser) parseMul() ast.Node {
	left := p.parsePow()
	for p.isOp("*") || p.isOp("/") {
		op := p.advance().Value
		right := p.parsePow()
		left = &ast.BinOp{Op: op, Left: left, Right: right}
	}
	return left
}

func (p *Parser) parsePow() ast.Node {
	left := p.parsePrimary()
	if p.isOp("^") {
		p.advance()
		right := p.parsePrimary()
		left = &ast.BinOp{Op: "^", Left: left, Right: right}
	}
	return left
}

func (p *Parser) parsePrimary() ast.Node {
	switch p.currentType() {
	case TokenNumber:
		val, _ := strconv.ParseFloat(p.advance().Value, 64)
		return &ast.Literal{Value: val}
	case TokenString:
		return &ast.StrLit{Value: p.advance().Value}
	case TokenIdent:
		return p.parseAssignOrCall()
	case TokenSymbol:
		if p.currentValue() == "(" {
			p.advance()
			expr := p.parseExpr()
			p.expectSym(")")
			return expr
		}
	}
	panic(fmt.Sprintf("Unexpected token: %v", p.current()))
}

func (p *Parser) parseArgs() []ast.Node {
	p.expectSym("(")
	var args []ast.Node
	for p.currentValue() != ")" {
		args = append(args, p.parseExpr())
		if p.currentValue() == "," {
			p.advance()
		}
	}
	p.expectSym(")")
	return args
}

func (p *Parser) current() TokenType {
	return p.tokens[p.pos].Type
}

func (p *Parser) currentType() TokenType {
	return p.tokens[p.pos].Type
}

func (p *Parser) currentValue() string {
	return p.tokens[p.pos].Value
}

func (p *Parser) advance() Token {
	tok := p.tokens[p.pos]
	p.pos++
	return tok
}

func (p *Parser) expect(tt TokenType) Token {
	if p.currentType() != tt {
		panic(fmt.Sprintf("Expected %v, got %v", tt, p.currentType()))
	}
	return p.advance()
}

func (p *Parser) expectOp(op string) {
	if !p.isOp(op) {
		panic(fmt.Sprintf("Expected op %s, got %s", op, p.currentValue()))
	}
	p.advance()
}

func (p *Parser) expectSym(sym string) {
	if p.currentValue() != sym {
		panic(fmt.Sprintf("Expected sym %s, got %s", sym, p.currentValue()))
	}
	p.advance()
}

func (p *Parser) expectKeyword(kw string) {
	if p.currentValue() != kw {
		panic(fmt.Sprintf("Expected keyword %s, got %s", kw, p.currentValue()))
	}
	p.advance()
}

func (p *Parser) isOp(op string) bool {
	return p.currentType() == TokenOperator && p.currentValue() == op
}

func isLetter(c byte) bool {
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_'
}

func isDigit(c byte) bool {
	return c >= '0' && c <= '9'
}

File: compiler/src/parser.rs
// src/parser.rs - Parser implementation

use crate::ast::Ast;

#[derive(Debug, Clone)]
enum Token {
    Ident(String),
    Number(f64),
    StringLit(String),
    Operator(String),
    Keyword(String),
    Symbol(String),
    Eof,
}

pub struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser {
    pub fn new(code: &str) -> Self {
        // Tokenization logic (expanded)
        let mut tokens = Vec::new();
        let mut chars = code.chars().peekable();
        while chars.peek().is_some() {
            let ch = *chars.peek().unwrap();
            match ch {
                ' ' | '\t' | '\n' | '\r' => { chars.next(); continue; }
                'a'..='z' | 'A'..='Z' | '_' => {
                    let mut id = String::new();
                    while let Some(&c) = chars.peek() {
                        if c.is_alphanumeric() || c == '_' {
                            id.push(c);
                            chars.next();
                        } else {
                            break;
                        }
                    }
                    if ["if", "else", "while", "for", "fn", "var", "write"].contains(&id.as_str()) {
                        tokens.push(Token::Keyword(id));
                    } else {
                        tokens.push(Token::Ident(id));
                    }
                }
                '0'..='9' | '.' => {
                    let mut num_str = String::new();
                    let mut has_dot = false;
                    while let Some(&c) = chars.peek() {
                        if c.is_digit(10) {
                            num_str.push(c);
                            chars.next();
                        } else if c == '.' && !has_dot {
                            num_str.push(c);
                            has_dot = true;
                            chars.next();
                        } else {
                            break;
                        }
                    }
                    tokens.push(Token::Number(num_str.parse().unwrap_or(0.0)));
                }
                '"' => {
                    chars.next();
                    let mut s = String::new();
                    while let Some(&c) = chars.peek() {
                        if c == '"' {
                            chars.next();
                            break;
                        }
                        s.push(c);
                        chars.next();
                    }
                    tokens.push(Token::StringLit(s));
                }
                '+' | '-' | '*' | '/' | '^' | '=' | '<' | '>' | '!' | '&' | '|' | '[' | ']' | '(' | ')' | '{' | '}' | ':' | ';' | ',' => {
                    let op = chars.next().unwrap().to_string();
                    tokens.push(if "+-*/^=<>!&|".contains(&op) { Token::Operator(op) } else { Token::Symbol(op) });
                }
                '@' => {
                    // Single line comment
                    chars.next();
                    while let Some(&c) = chars.peek() {
                        if c == '\n' { break; }
                        chars.next();
                    }
                }
                '!' => {
                    // Multi line comment
                    chars.next();
                    let mut depth = 1;
                    while depth > 0 && chars.peek().is_some() {
                        let c = chars.next().unwrap();
                        if c == '!' { depth -= 1; }
                    }
                }
                _ => { chars.next(); } // Ignore unknown
            }
        }
        tokens.push(Token::Eof);
        Parser { tokens, pos: 0 }
    }

    pub fn parse(&mut self) -> Vec<Ast> {
        let mut stmts = Vec::new();
        while self.pos < self.tokens.len() - 1 {
            stmts.push(self.parse_stmt());
        }
        stmts
    }

    fn parse_stmt(&mut self) -> Ast {
        match &self.peek() {
            Token::Keyword(k) if k == "var" => self.parse_var_decl(),
            Token::Keyword(k) if k == "fn" => self.parse_func_def(),
            Token::Keyword(k) if k == "if" => self.parse_if(),
            Token::Keyword(k) if k == "while" => self.parse_while(),
            Token::Keyword(k) if k == "for" => self.parse_for(),
            Token::Keyword(k) if k == "write" => self.parse_write(),
            Token::Ident(_) => self.parse_assign_or_call(),
            _ => self.parse_expr(),
        }
    }

    fn parse_var_decl(&mut self) -> Ast {
        self.next(); // var
        let name = if let Token::Ident(n) = self.next() { n } else { panic!("Expected ident"); };
        if let Token::Operator(op) = self.peek() {
            if op == "=" {
                self.next();
                let value = self.parse_expr();
                Ast::VarDecl(name, Box::new(value))
            } else {
                panic!("Expected =");
            }
        } else {
            panic!("Expected =");
        }
    }

    fn parse_assign_or_call(&mut self) -> Ast {
        let name = if let Token::Ident(n) = self.next() { n } else { unreachable!() };
        match self.peek() {
            Token::Symbol(s) if s == "(" => Ast::FuncCall(name, self.parse_args()),
            Token::Operator(op) if op == "=" => {
                self.next();
                Ast::Assign(name, Box::new(self.parse_expr()))
            }
            Token::Symbol(s) if s == "[" => {
                self.next();
                let index = self.parse_expr();
                self.expect_symbol("]");
                Ast::Index(name, Box::new(index))
            }
            _ => Ast::Var(name),
        }
    }

    fn parse_func_def(&mut self) -> Ast {
        self.next(); // fn
        let name = if let Token::Ident(n) = self.next() { n } else { panic!("Expected func name"); };
        self.expect_symbol("(");
        let mut params = Vec::new();
        while !matches!(&self.peek(), Token::Symbol(s) if s == ")") {
            if let Token::Ident(p) = self.next() { params.push(p); }
            if matches!(&self.peek(), Token::Symbol(s) if s == ",") { self.next(); }
        }
        self.expect_symbol(")");
        self.expect_symbol("{");
        let body = self.parse_block();
        self.expect_symbol("}");
        Ast::FuncDef(name, params, body)
    }

    fn parse_if(&mut self) -> Ast {
        self.next(); // if
        let cond = self.parse_expr();
        self.expect_symbol("{");
        let then = self.parse_block();
        self.expect_symbol("}");
        let els = if matches!(&self.peek(), Token::Keyword(k) if k == "else") {
            self.next();
            self.expect_symbol("{");
            let e = self.parse_block();
            self.expect_symbol("}");
            Some(e)
        } else {
            None
        };
        Ast::If(Box::new(cond), then, els)
    }

    fn parse_while(&mut self) -> Ast {
        self.next(); // while
        let cond = self.parse_expr();
        self.expect_symbol("{");
        let body = self.parse_block();
        self.expect_symbol("}");
        Ast::While(Box::new(cond), body)
    }

    fn parse_for(&mut self) -> Ast {
        self.next(); // for
        let var = if let Token::Ident(v) = self.next() { v } else { panic!("Expected var"); };
        self.expect_keyword("in");
        let start = self.parse_expr();
        self.expect_operator("..");
        let end = self.parse_expr();
        self.expect_symbol("{");
        let body = self.parse_block();
        self.expect_symbol("}");
        Ast::For(var, Box::new(start), Box::new(end), body)
    }

    fn parse_write(&mut self) -> Ast {
        self.next(); // write
        Ast::FuncCall("write".to_string(), vec![self.parse_expr()])
    }

    fn parse_block(&mut self) -> Vec<Ast> {
        let mut block = Vec::new();
        while !matches!(&self.peek(), Token::Symbol(s) if s == "}") && !matches!(&self.peek(), Token::Eof) {
            block.push(self.parse_stmt());
        }
        block
    }

    fn parse_expr(&mut self) -> Ast {
        self.parse_add()
    }

    fn parse_add(&mut self) -> Ast {
        let mut left = self.parse_mul();
        while matches!(&self.peek(), Token::Operator(op) if ["+", "-"].contains(&op.as_str())) {
            let op = self.next_operator();
            let right = self.parse_mul();
            left = Ast::BinOp(op, Box::new(left), Box::new(right));
        }
        left
    }

    fn parse_mul(&mut self) -> Ast {
        let mut left = self.parse_pow();
        while matches!(&self.peek(), Token::Operator(op) if ["*", "/"].contains(&op.as_str())) {
            let op = self.next_operator();
            let right = self.parse_pow();
            left = Ast::BinOp(op, Box::new(left), Box::new(right));
        }
        left
    }

    fn parse_pow(&mut self) -> Ast {
        let mut left = self.parse_primary();
        while matches!(&self.peek(), Token::Operator(op) if op == "^") {
            self.next();
            let right = self.parse_primary();
            left = Ast::BinOp("^".to_string(), Box::new(left), Box::new(right));
        }
        left
    }

    fn parse_primary(&mut self) -> Ast {
        match self.peek().clone() {
            Token::Number(n) => { self.next(); Ast::Literal(n) }
            Token::StringLit(s) => { self.next(); Ast::StrLit(s) }
            Token::Ident(id) => self.parse_assign_or_call(),
            Token::Symbol(s) if s == "(" => {
                self.next();
                let expr = self.parse_expr();
                self.expect_symbol(")");
                expr
            }
            Token::Symbol(s) if s == "[" => self.parse_array(),
            _ => panic!("Unexpected token: {:?}", self.peek()),
        }
    }

    fn parse_array(&mut self) -> Ast {
        self.next(); // [
        let mut elements = Vec::new();
        while !matches!(&self.peek(), Token::Symbol(s) if s == "]") {
            elements.push(self.parse_expr());
            if matches!(&self.peek(), Token::Symbol(s) if s == ",") { self.next(); }
        }
        self.expect_symbol("]");
        Ast::Array(elements)
    }

    fn parse_args(&mut self) -> Vec<Ast> {
        self.expect_symbol("(");
        let mut args = Vec::new();
        while !matches!(&self.peek(), Token::Symbol(s) if s == ")") {
            args.push(self.parse_expr());
            if matches!(&self.peek(), Token::Symbol(s) if s == ",") { self.next(); }
        }
        self.expect_symbol(")");
        args
    }

    fn next(&mut self) -> Token {
        let tok = self.tokens[self.pos].clone();
        self.pos += 1;
        tok
    }

    fn peek(&self) -> Token {
        self.tokens[self.pos].clone()
    }

    fn expect_symbol(&mut self, sym: &str) {
        if matches!(&self.peek(), Token::Symbol(s) if s == sym) {
            self.next();
        } else {
            panic!("Expected symbol {}, got {:?}", sym, self.peek());
        }
    }

    fn expect_operator(&mut self, op: &str) {
        if matches!(&self.peek(), Token::Operator(o) if o == op) {
            self.next();
        } else {
            panic!("Expected operator {}, got {:?}", op, self.peek());
        }
    }

    fn expect_keyword(&mut self, kw: &str) {
        if matches!(&self.peek(), Token::Keyword(k) if k == kw) {
            self.next();
        } else {
            panic!("Expected keyword {}, got {:?}", kw, self.peek());
        }
    }

    fn next_operator(&mut self) -> String {
        if let Token::Operator(op) = self.next() { op } else { panic!("Expected operator"); }
    }
}
